nifi-nar-bundles/nifi-email-bundle/nifi-email-processors/src/main/java/org/apache/nifi/processors/email/ExtractEmailHeaders.java

```
@@ -0,0 +1,234 @@
+        relationships.add(REL_SUCCESS);
+        relationships.add(REL_FAILURE);
+        this.relationships = Collections.unmodifiableSet(relationships);
+
+        final List<PropertyDescriptor> descriptors = new ArrayList<>();
+
+        descriptors.add(CAPTURED_HEADERS);
+        this.descriptors = Collections.unmodifiableList(descriptors);
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) {
+        final ComponentLog logger = getLogger();
+        final Map<String, String> attributes = new HashMap<>();
+        final FlowFile originalFlowFile = session.get();
+        if (originalFlowFile == null) {
+            return;
+        }
+
+
+        final List<String> capturedHeadersList = Arrays.asList(context.getProperty(CAPTURED_HEADERS).getValue().toLowerCase().split(":"));
+
+
+        session.read(originalFlowFile, new InputStreamCallback() {
+                @Override
+                public void process(final InputStream rawIn) throws IOException {
+                    try (final InputStream in = new BufferedInputStream(rawIn)) {
+                        Properties props = new Properties();
+                        Session mailSession = Session.getDefaultInstance(props, null);
+                        MimeMessage originalMessage = new MimeMessage(mailSession, in);
+                        MimeMessageParser parser = new MimeMessageParser(originalMessage).parse();
+                        // RFC-2822 determines that a message must have a "From:" header
+                        // if a message lacks the field, it is flagged as invalid
+                        Address[] from = originalMessage.getFrom();
+                        Date sentDate = originalMessage.getSentDate();
+                        if (from == null || sentDate == null ) {
+                            // See ya later my friend...
+                            throw new MessagingException("Message failed RFC2822 validation");
```

The MessageException is caught and if number of extracted attributes is 0, the original flow file will be sent to FAILURE. However, the exception is swallowed and its stack trace will not be logged, it will be hard to debug if something unexpected happens other than MessageException. At least, there's a possibility that `new MimeMessageParser(originalMessage).parse()` throws Exception. To make investigation easy, logging Exception's stack trace is highly recommended, by passing the exception as the last element in logger.error method parameter array.